Headless conjure up failed at juju_wait

## Report

Thank you for trying **conjure-up**! Before reporting a bug please make sure you've gone through this checklist:

- [X] Is this problem already documented at https://docs.ubuntu.com/conjure-up/en/troubleshoot#common-spell-problems
- [X] Is conjure-up running inside a virtual machine?
- [X] Is this problem reproducible with `sudo snap refresh conjure-up --edge`?

## Please provide the output of the following commands

```
which juju
/snap/bin/juju
juju version
2.3-rc2-xenial-amd64

which conjure-up
/snap/bin/conjure-up
conjure-up --version
conjure-up 2.5-alpha1

which lxc
/usr/bin/lxc
/snap/bin/lxc config show
- no such file or directory
/snap/bin/lxc version
- no such file or directory

cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.3 LTS"
```

Please attach tarball of **~/.cache/conjure-up**:
[conjure-up.tar.gz](https://github.com/conjure-up/conjure-up/files/1500225/conjure-up.tar.gz)
```
tar cvzf conjure-up.tar.gz ~/.cache/conjure-up
```

## Sosreport

Please attach a sosreport:

```
sudo apt install sosreport
sosreport


sosreport (version 3.4)

no valid plugins were enabled
```

The resulting output file can be attached to this issue.


## What Spell was Selected?
canonical-kubernetes

## What provider (aws, maas, localhost, etc)?
aws 

### MAAS Users
Which version of MAAS?

## Commands ran
Please outline what commands were run to install and execute conjure-up:
```
$ aws ec2 run-instances --image-id ami-da05a4a0 --key-name redacted --security-group-ids sg-redacted --instance-type t2.small --subnet-id subnet-redacted
$ ansible-playbook -i '172.16.204.133,' bootstrapUbuntu.yml --private-key ~/.ssh/redacted
--> install python2 (required for ansible) add users + keys, install core tools such as jq, nmap and atop
$ ssh 172.16.204.133
$ sudo snap install juju --classic
$ sudo snap refresh juju --edge
$ sudo snap install conjure-up --classic
$ sudo snap refresh conjure-up --edge
$ export AWS_ACCESS_KEY_ID=<personal access key>
$ export AWS_SECRET_ACCESS_KEY=<personal secret>
$ juju bootstrap aws/us-east-1 pre-test-k8s --config vpc-id=vpc-redacted --config vpc-id-force=true --to "subnet=subnet-redacted" --constraints "root-disk=128G mem=4G" --bootstrap-series xenial
$ juju add-credential aws
--> add kubernetes IAM user credentials
$ conjure-up canonical-kubernetes aws/us-east-1 pre-test-k8s
[info] Summoning canonical-kubernetes to aws
[info] Creating Juju model.
[info] Juju model created.
[info] Running step: 00_process-providertype before-deploy.
[info] Deploying etcd...
[info] Deploying kubernetes-master...
[info] Deploying easyrsa...
[info] Deploying kubernetes-worker...
[info] Deploying flannel...
[info] Deploying kubeapi-load-balancer...
[info] easyrsa: deployed, installing.
[info] flannel: deployed, installing.
[info] kubernetes-master: deployed, installing.
[info] Exposing kubeapi-load-balancer.
[info] kubeapi-load-balancer: deployed, installing.
[info] etcd: deployed, installing.
[info] Setting relation etcd:db <-> flannel:etcd
[info] Exposing kubernetes-worker.
[info] kubernetes-worker: deployed, installing.
[info] Setting relation flannel:cni <-> kubernetes-master:cni
[info] Setting relation kubeapi-load-balancer:apiserver <-> kubernetes-master:kube-api-endpoint
[info] Setting relation easyrsa:client <-> kubernetes-master:certificates
[info] Setting relation flannel:cni <-> kubernetes-worker:cni
[info] Setting relation kubeapi-load-balancer:loadbalancer <-> kubernetes-master:loadbalancer
[info] Setting relation easyrsa:client <-> etcd:certificates
[info] Setting relation kubernetes-master:kube-control <-> kubernetes-worker:kube-control
[info] Setting relation easyrsa:client <-> kubeapi-load-balancer:certificates
[info] Setting relation kubeapi-load-balancer:website <-> kubernetes-worker:kube-api-endpoint
[info] Setting relation etcd:db <-> kubernetes-master:etcd
[info] Setting relation easyrsa:client <-> kubernetes-worker:certificates
[info] Waiting for deployment to settle.
[error] Some applications failed to start successfully.
[warning] Shutting down
```


## Additional Information

Piling on here to say that I've hit the same issue.  In my case, I was using conjure-up 2.4.1-20171121.0211 with juju 2.3-rc2+develop-fbb8949, but the results are the same:

```
[error] Some applications failed to start successfully.
```

And the root cause in conjure-up.log from both @ryan-peachey-seon and my logs is a failure to find juju_wait.  Here's the relevant bits from @ryan-peachey-seon's log:

```
2017-11-23 21:19:07,301 [INFO] conjure-up/canonical-kubernetes - common.py:56 - Waiting for deployment to settle.
2017-11-23 21:19:07,465 [ERROR] conjure-up/canonical-kubernetes - juju.py:894 - Traceback (most recent call last):
  File "/snap/conjure-up/842/bin/juju-wait", line 7, in <module>
    from juju_wait import wait_cmd
ModuleNotFoundError: No module named 'juju_wait'
2017-11-23 21:19:07,467 [DEBUG] conjure-up/canonical-kubernetes - events.py:53 - Setting Error at conjureup/events.py:148
2017-11-23 21:19:07,467 [ERROR] conjure-up/canonical-kubernetes - events.py:160 - Unhandled exception in <Task finished coro=<DeployController._wait_for_applications() done, defined at /snap/conjure-up/842/lib/python3.6/site-packages/conjureup/controllers/deploy/tui.py:12> exception=DeploymentFailure('Some applications failed to start successfully.',)>
Traceback (most recent call last):
  File "/snap/conjure-up/842/lib/python3.6/site-packages/conjureup/controllers/deploy/tui.py", line 13, in _wait_for_applications
    await common.wait_for_applications(utils.info)
  File "/snap/conjure-up/842/lib/python3.6/site-packages/conjureup/controllers/deploy/common.py", line 59, in wait_for_applications
    await juju.wait_for_deployment()
  File "/snap/conjure-up/842/lib/python3.6/site-packages/conjureup/juju.py", line 896, in wait_for_deployment
    "Some applications failed to start successfully.")
conjureup.errors.DeploymentFailure: Some applications failed to start successfully.
```

And here's mine:

```
2017-11-27 13:26:53,270 [DEBUG] conjure-up/hadoop-processing - events.py:53 - Setting DeploymentComplete at conjureup/controllers/deploy/common.py:46
2017-11-27 13:26:53,270 [DEBUG] conjure-up/hadoop-processing - events.py:53 - Received DeploymentComplete at conjureup/controllers/deploy/common.py:50 in task _wait_for_applications at conjureup/controllers/deploy/gui.py:90
2017-11-27 13:26:53,270 [INFO] conjure-up/hadoop-processing - common.py:52 - Waiting for deployment to settle.
2017-11-27 13:26:53,537 [ERROR] conjure-up/hadoop-processing - juju.py:894 - Traceback (most recent call last):
  File "/snap/conjure-up/837/bin/juju-wait", line 7, in <module>
    from juju_wait import wait_cmd
ModuleNotFoundError: No module named 'juju_wait'
```

Also seeing same 'ModuleNotFoundError: No module named 'juju_wait'' error when deploying nova-lxd on localhost. 

snap list
Name        Version              Rev   Developer  Notes
conjure-up  2.4.1-20171121.0211  837   canonical  classic
core        16-2.29.3            3440  canonical  core
juju        2.2.6                2739  canonical  classic
lxd         2.20                 5061  canonical  -

lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.3 LTS
Release:        16.04
Codename:       xenial

Tried the conjure up edge channel with the same result.

```
2017-11-28 00:59:49,276 [DEBUG] conjure-up/openstack-novalxd - events.py:53 - Received DeploymentComplete at conjureup/controllers/deploy/common.py:50 in task _wait_for_applications at conjureup/controllers/deploy/gui.py:90
2017-11-28 00:59:49,276 [INFO] conjure-up/openstack-novalxd - common.py:52 - Waiting for deployment to settle.
2017-11-28 00:59:49,563 [ERROR] conjure-up/openstack-novalxd - juju.py:894 - Traceback (most recent call last):
  File "/snap/conjure-up/837/bin/juju-wait", line 7, in <module>
    from juju_wait import wait_cmd
ModuleNotFoundError: No module named 'juju_wait'
2017-11-28 00:59:49,565 [DEBUG] conjure-up/openstack-novalxd - events.py:53 - Setting Error at conjureup/events.py:148
2017-11-28 00:59:49,566 [ERROR] conjure-up/openstack-novalxd - events.py:160 - Unhandled exception in <Task finished coro=<DeployController._wait_for_applications() done, defined at /snap/conjure-up/837/lib/python3.6/site-packages/conjureup/controllers/deploy/gui.py:89> exception=DeploymentFailure('Some applications failed to start successfully.',)>
Traceback (most recent call last):
  File "/snap/conjure-up/837/lib/python3.6/site-packages/conjureup/controllers/deploy/gui.py", line 90, in _wait_for_applications
    await common.wait_for_applications(app.ui.set_footer)
  File "/snap/conjure-up/837/lib/python3.6/site-packages/conjureup/controllers/deploy/common.py", line 55, in wait_for_applications
    await juju.wait_for_deployment()
  File "/snap/conjure-up/837/lib/python3.6/site-packages/conjureup/juju.py", line 896, in wait_for_deployment
    "Some applications failed to start successfully.")
conjureup.errors.DeploymentFailure: Some applications failed to start successfully.
2017-11-28 00:59:49,578 [DEBUG] conjure-up/openstack-novalxd - __init__.py:21 - Showing dialog for exception: Some applications failed to start successfully.
2017-11-28 01:00:24,802 [DEBUG] conjure-up/openstack-novalxd - events.py:53 - Setting Shutdown at conjureup/events.py:144
2017-11-28 01:00:24,803 [DEBUG] conjure-up/openstack-novalxd - events.py:53 - Received Shutdown at conjureup/events.py:175
2017-11-28 01:00:24,803 [INFO] conjure-up/openstack-novalxd - events.py:179 - Shutting down
2017-11-28 01:00:24,803 [INFO] conjure-up/openstack-novalxd - app_config.py:190 - Storing conjure-up state
2017-11-28 01:00:24,851 [INFO] conjure-up/openstack-novalxd - app_config.py:194 - State saved in model config
2017-11-28 01:00:24,853 [INFO] conjure-up/openstack-novalxd - events.py:189 - Disconnecting model
2017-11-28 01:00:24,859 [INFO] conjure-up/openstack-novalxd - events.py:191 - Disconnected
2017-11-28 01:00:24,860 [DEBUG] conjure-up/openstack-novalxd - events.py:201 - Cancelling pending task: <Task finished coro=<WebSocketCommonProtocol.close_connection() done, defined at /snap/conjure-up/837/lib/python3.6/site-packages/websockets/protocol.py:695> result=None>
2017-11-28 01:00:24,861 [DEBUG] conjure-up/openstack-novalxd - events.py:201 - Cancelling pending task: <Task finished coro=<WebSocketCommonProtocol.transfer_data() done, defined at /snap/conjure-up/837/lib/python3.6/site-packages/websockets/protocol.py:486> result=None>```

--- a/Makefile
+++ b/Makefile
@@ -25,6 +25,7 @@ travis-sysdeps:
 	@sudo pip install tox
 	@sudo snap install juju --classic --edge
 	@sudo snap refresh lxd --edge
+	@sudo snap install juju-wait --classic
 
 .PHONY: install
 install: snap

--- a/conjureup/app.py
+++ b/conjureup/app.py
@@ -199,10 +199,6 @@ def main():
         print("")
         sys.exit(1)
 
-    # Make sure juju paths are setup
-    juju.set_bin_path()
-    juju.set_wait_path()
-
     # Verify we can access ~/.local/share/juju if it exists
     juju_dir = pathlib.Path('~/.local/share/juju').expanduser()
     if juju_dir.exists():
@@ -236,6 +232,10 @@ def main():
     app.log = setup_logging(app,
                             os.path.join(opts.cache_dir, 'conjure-up.log'),
                             opts.debug)
+
+    # Make sure juju paths are setup
+    juju.set_bin_path()
+    juju.set_wait_path()
 
     if app.argv.conf_file.expanduser().exists():
         conf = configparser.ConfigParser()

--- a/conjureup/app.py
+++ b/conjureup/app.py
@@ -199,6 +199,10 @@ def main():
         print("")
         sys.exit(1)
 
+    # Make sure juju paths are setup
+    juju.set_bin_path()
+    juju.set_wait_path()
+
     # Verify we can access ~/.local/share/juju if it exists
     juju_dir = pathlib.Path('~/.local/share/juju').expanduser()
     if juju_dir.exists():

--- a/conjureup/app_config.py
+++ b/conjureup/app_config.py
@@ -3,6 +3,11 @@
 import json
 from types import SimpleNamespace
 
+
+class AppConfigAttributeError(Exception):
+    pass
+
+
 bootstrap = SimpleNamespace(
     # Is bootstrap running
     running=False,

--- a/conjureup/app_config.py
+++ b/conjureup/app_config.py
@@ -27,7 +27,13 @@ juju = SimpleNamespace(
     client=None,
 
     # Is authenticated?
-    authenticated=False
+    authenticated=False,
+
+    # Path to juju binary
+    bin_path=None,
+
+    # Path to juju-wait binary
+    wait_path=None
 )
 
 

--- a/conjureup/app_config.py
+++ b/conjureup/app_config.py
@@ -3,11 +3,6 @@
 import json
 from types import SimpleNamespace
 
-
-class AppConfigAttributeError(Exception):
-    pass
-
-
 bootstrap = SimpleNamespace(
     # Is bootstrap running
     running=False,

--- a/conjureup/errors.py
+++ b/conjureup/errors.py
@@ -12,3 +12,11 @@ class ControllerNotFoundException(Exception):
 
 class DeploymentFailure(Exception):
     "A failure from a deployed model"
+
+
+class JujuBinaryNotFound(Exception):
+    "A failure finding juju or juju-wait executable"
+
+
+class AppConfigAttributeError(Exception):
+    "A failure to lookup attribute in app_config object"

--- a/conjureup/juju.py
+++ b/conjureup/juju.py
@@ -14,20 +14,12 @@ from bundleplacer.charmstore_api import CharmStoreID
 from juju.model import Model
 
 from conjureup import consts, errors, events, utils
-from conjureup.app_config import AppConfigAttributeError, app
+from conjureup.app_config import app
 from conjureup.utils import is_linux, juju_path, run, spew
 
 JUJU_ASYNC_QUEUE = "juju-async-queue"
 
 PENDING_DEPLOYS = 0
-
-
-class JujuBinaryNotFound(Exception):
-    pass
-
-
-class JujuAttributeError(Exception):
-    pass
 
 
 def _check_bin_candidates(candidates, bin_property):
@@ -38,7 +30,7 @@ def _check_bin_candidates(candidates, bin_property):
     # we don't use $PATH because we have definite preferences which one we use
     # and we don't want to leave it up to the user
     if not hasattr(app.juju, bin_property):
-        raise AppConfigAttributeError(
+        raise errors.AppConfigAttributeError(
             "Unknown juju property: {}".format(bin_property))
     for candidate in candidates:
         if os.access(candidate, os.X_OK):
@@ -46,7 +38,7 @@ def _check_bin_candidates(candidates, bin_property):
             app.log.debug("{} candidate found".format(bin_property))
             break
     else:
-        raise JujuBinaryNotFound(
+        raise errors.JujuBinaryNotFound(
             "Unable to locate a candidate executable for {}.".format(
                 candidates))
 

--- a/conjureup/juju.py
+++ b/conjureup/juju.py
@@ -14,7 +14,7 @@ from bundleplacer.charmstore_api import CharmStoreID
 from juju.model import Model
 
 from conjureup import consts, errors, events, utils
-from conjureup.app_config import app
+from conjureup.app_config import AppConfigAttributeError, app
 from conjureup.utils import is_linux, juju_path, run, spew
 
 JUJU_ASYNC_QUEUE = "juju-async-queue"
@@ -26,6 +26,10 @@ class JujuBinaryNotFound(Exception):
     pass
 
 
+class JujuAttributeError(Exception):
+    pass
+
+
 def _check_bin_candidates(candidates, bin_property):
     """ Checks a list of binary paths to verify they exist and are
     executable
@@ -33,9 +37,12 @@ def _check_bin_candidates(candidates, bin_property):
     # search candidate paths, in order, for the binary (ie juju, juju-wait)
     # we don't use $PATH because we have definite preferences which one we use
     # and we don't want to leave it up to the user
+    if not hasattr(app.juju, bin_property):
+        raise AppConfigAttributeError(
+            "Unknown juju property: {}".format(bin_property))
     for candidate in candidates:
         if os.access(candidate, os.X_OK):
-            bin_property = candidate
+            setattr(app.juju, bin_property, candidate)
             app.log.debug("{} candidate found".format(bin_property))
             break
     else:
@@ -53,7 +60,7 @@ def set_bin_path():
         '/usr/bin/juju',
         '/usr/local/bin/juju',
     ]
-    _check_bin_candidates(candidates, app.juju.bin_path)
+    _check_bin_candidates(candidates, 'bin_path')
     # Update $PATH so that we make sure this candidate is used
     # first.
     app.env['PATH'] = "{}:{}".format(Path(app.juju.bin_path).parent,
@@ -70,7 +77,7 @@ def set_wait_path():
         '/usr/local/bin/juju-wait',
     ]
 
-    _check_bin_candidates(candidates, app.juju.wait_path)
+    _check_bin_candidates(candidates, 'wait_path')
 
 
 def read_config(name):

--- a/conjureup/juju.py
+++ b/conjureup/juju.py
@@ -22,6 +22,38 @@ JUJU_ASYNC_QUEUE = "juju-async-queue"
 PENDING_DEPLOYS = 0
 
 
+class JujuBinaryNotFound(Exception):
+    pass
+
+
+def set_bin_path():
+    """ Sets the juju binary path
+    """
+    if utils.is_darwin():
+        app.juju.bin_path = '/usr/local/bin/juju'
+    elif Path('/snap/bin/juju').exists():
+        app.juju.bin_path = '/snap/bin/juju'
+    elif Path('/snap/bin/conjure-up.juju').exists():
+        app.juju.bin_path = '/snap/bin/conjure-up.juju'
+    else:
+        raise JujuBinaryNotFound("Unable to locate a juju binary.")
+    app.log.debug("juju binary path: {}".format(app.juju.bin_path))
+
+
+def set_wait_path():
+    """ Sets juju-wait path
+    """
+    if utils.is_darwin():
+        app.juju.wait_path = '/usr/local/bin/juju-wait'
+    elif Path('/snap/bin/juju-wait').exists():
+        app.juju.wait_path = '/snap/bin/juju-wait'
+    elif Path('/snap/bin/conjure-up.juju-wait').exists():
+        app.juju.wait_path = '/snap/bin/conjure-up.juju-wait'
+    else:
+        raise JujuBinaryNotFound("Unable to locate a juju-wait binary.")
+    app.log.debug("juju-wait binary path: {}".format(app.juju.wait_path))
+
+
 def read_config(name):
     """ Reads a juju config file
 
@@ -136,7 +168,8 @@ async def bootstrap(controller, cloud, model='conjure-up', series="xenial",
         app.log.debug("Bootstrapping to set region: {}")
         cloud = "{}/{}".format(app.provider.cloud, app.provider.region)
 
-    cmd = ["juju", "bootstrap", cloud, controller, "--default-model", model]
+    cmd = [app.juju.bin_path, "bootstrap",
+           cloud, controller, "--default-model", model]
 
     def add_config(k, v):
         cmd.extend(["--config", "{}={}".format(k, v)])
@@ -250,7 +283,8 @@ def autoload_credentials():
     """ Automatically checks known places for cloud credentials
     """
     try:
-        run('juju autoload-credentials', shell=True, check=True)
+        run('{} autoload-credentials'.format(
+            app.juju.bin_path), shell=True, check=True)
     except CalledProcessError:
         return False
     return True
@@ -293,7 +327,7 @@ def get_credentials(secrets=True):
     Returns:
     List of credentials
     """
-    cmd = 'juju list-credentials --format yaml'
+    cmd = '{} list-credentials --format yaml'.format(app.juju.bin_path)
     if secrets:
         cmd += ' --show-secrets'
     sh = run(cmd, shell=True, stdout=PIPE, stderr=PIPE)
@@ -318,7 +352,8 @@ def get_regions(cloud):
     Returns:
     Dictionary of all known regions for cloud
     """
-    sh = run('juju list-regions {} --format yaml'.format(cloud),
+    sh = run('{} list-regions {} --format yaml'.format(app.juju.bin_path,
+                                                       cloud),
              shell=True, stdout=PIPE, stderr=PIPE)
     stdout = sh.stdout.decode('utf8')
     stderr = sh.stderr.decode('utf8')
@@ -341,7 +376,7 @@ def get_clouds():
     Returns:
     Dictionary of all known clouds including newly created MAAS/Local
     """
-    sh = run('juju list-clouds --format yaml',
+    sh = run('{} list-clouds --format yaml'.format(app.juju.bin_path),
              shell=True, stdout=PIPE, stderr=PIPE)
     if sh.returncode > 0:
         raise Exception(
@@ -457,7 +492,8 @@ def add_cloud(name, config):
                             delete=False) as tempf:
         output = yaml.safe_dump(_config, default_flow_style=False)
         spew(tempf.name, output)
-        sh = run('juju add-cloud {} {}'.format(name, tempf.name),
+        sh = run('{} add-cloud {} {}'.format(app.juju.bin_path,
+                                             name, tempf.name),
                  shell=True, stdout=PIPE, stderr=PIPE)
         if sh.returncode > 0:
             raise Exception(
@@ -518,7 +554,8 @@ def deploy(bundle):
             charmstore path.
     """
     try:
-        return run('juju deploy {}'.format(bundle), shell=True,
+        return run('{} deploy {}'.format(app.juju.bin_path,
+                                         bundle), shell=True,
                    stdout=DEVNULL, stderr=PIPE)
     except CalledProcessError as e:
         raise e
@@ -700,7 +737,8 @@ def get_controller_info(name=None):
     Arguments:
     name: if set shows info controller, otherwise displays current.
     """
-    cmd = 'juju show-controller --format yaml'
+    cmd = '{} show-controller --format yaml'.format(
+        app.juju.bin_path)
     if name is not None:
         cmd += ' {}'.format(name)
     sh = run(cmd, shell=True, stdout=PIPE, stderr=PIPE)
@@ -722,8 +760,9 @@ def get_controllers():
     Returns:
     List of known controllers
     """
-    sh = run('juju list-controllers --format yaml',
-             shell=True, stdout=PIPE, stderr=PIPE)
+    sh = run('{} list-controllers --format yaml'.format(
+        app.juju.bin_path),
+        shell=True, stdout=PIPE, stderr=PIPE)
     if sh.returncode > 0:
         raise LookupError(
             "Unable to list controllers: {}".format(sh.stderr.decode('utf8')))
@@ -845,7 +884,8 @@ def get_models(controller):
     Returns:
     List of known models
     """
-    sh = run('juju list-models --format yaml -c {}'.format(controller),
+    sh = run('{} list-models --format yaml -c {}'.format(app.juju.bin_path,
+                                                         controller),
              shell=True, stdout=PIPE, stderr=PIPE)
     if sh.returncode > 0:
         raise LookupError(
@@ -864,7 +904,9 @@ def get_current_model():
 def version():
     """ Returns version of Juju
     """
-    sh = run('juju version', shell=True, stdout=PIPE, stderr=PIPE)
+    sh = run('{} version'.format(
+        app.juju.bin_path),
+        shell=True, stdout=PIPE, stderr=PIPE)
     if sh.returncode > 0:
         raise Exception(
             "Unable to get Juju Version".format(sh.stderr.decode('utf8')))
@@ -881,7 +923,7 @@ async def wait_for_deployment(retries=3):
     if 'CONJURE_UP_MODE' in app.env and app.env['CONJURE_UP_MODE'] == "test":
         retries = 0
 
-    cmd = ['juju', 'wait', "-r{}".format(retries),
+    cmd = [app.juju.wait_path, "-r{}".format(retries),
            "-vwm", "{}:{}".format(app.provider.controller,
                                   app.provider.model)]
 

--- a/conjureup/juju.py
+++ b/conjureup/juju.py
@@ -26,32 +26,51 @@ class JujuBinaryNotFound(Exception):
     pass
 
 
+def _check_bin_candidates(candidates, bin_property):
+    """ Checks a list of binary paths to verify they exist and are
+    executable
+    """
+    # search candidate paths, in order, for the binary (ie juju, juju-wait)
+    # we don't use $PATH because we have definite preferences which one we use
+    # and we don't want to leave it up to the user
+    for candidate in candidates:
+        if os.access(candidate, os.X_OK):
+            bin_property = candidate
+            app.log.debug("{} candidate found".format(bin_property))
+            break
+    else:
+        raise JujuBinaryNotFound(
+            "Unable to locate a candidate executable for {}.".format(
+                candidates))
+
+
 def set_bin_path():
     """ Sets the juju binary path
     """
-    if utils.is_darwin():
-        app.juju.bin_path = '/usr/local/bin/juju'
-    elif Path('/snap/bin/juju').exists():
-        app.juju.bin_path = '/snap/bin/juju'
-    elif Path('/snap/bin/conjure-up.juju').exists():
-        app.juju.bin_path = '/snap/bin/conjure-up.juju'
-    else:
-        raise JujuBinaryNotFound("Unable to locate a juju binary.")
-    app.log.debug("juju binary path: {}".format(app.juju.bin_path))
+    candidates = [
+        '/snap/bin/juju',
+        '/snap/bin/conjure-up.juju',
+        '/usr/bin/juju',
+        '/usr/local/bin/juju',
+    ]
+    _check_bin_candidates(candidates, app.juju.bin_path)
+    # Update $PATH so that we make sure this candidate is used
+    # first.
+    app.env['PATH'] = "{}:{}".format(Path(app.juju.bin_path).parent,
+                                     app.env['PATH'])
 
 
 def set_wait_path():
     """ Sets juju-wait path
     """
-    if utils.is_darwin():
-        app.juju.wait_path = '/usr/local/bin/juju-wait'
-    elif Path('/snap/bin/juju-wait').exists():
-        app.juju.wait_path = '/snap/bin/juju-wait'
-    elif Path('/snap/bin/conjure-up.juju-wait').exists():
-        app.juju.wait_path = '/snap/bin/conjure-up.juju-wait'
-    else:
-        raise JujuBinaryNotFound("Unable to locate a juju-wait binary.")
-    app.log.debug("juju-wait binary path: {}".format(app.juju.wait_path))
+    candidates = [
+        '/snap/bin/juju-wait',
+        '/snap/bin/conjure-up.juju-wait',
+        '/usr/bin/juju-wait',
+        '/usr/local/bin/juju-wait',
+    ]
+
+    _check_bin_candidates(candidates, app.juju.wait_path)
 
 
 def read_config(name):

--- a/conjureup/utils.py
+++ b/conjureup/utils.py
@@ -265,7 +265,7 @@ async def can_sudo(password=None):
 def juju_version():
     """ Get current Juju version
     """
-    cmd = run_script('juju version')
+    cmd = run_script('{} version'.format(app.juju.bin_path))
     if cmd.returncode == 0:
         return parse_version(cmd.stdout.decode().strip())
     else:

--- a/snap/snapcraft.yaml
+++ b/snap/snapcraft.yaml
@@ -18,6 +18,9 @@ apps:
     - conjure-down
   juju:
     command: wrappers/juju
+  juju-wait:
+    command: wrappers/juju-wait
+
 parts:
   conjure-up:
     source: .

--- a/snap/wrappers/juju-wait
+++ b/snap/wrappers/juju-wait
@@ -2,18 +2,7 @@
 
 
 # Make sure we access our python and juju binaries first
-debian_socket=/var/lib/lxd
-snap_lxd_socket=/var/snap/lxd/common/lxd
-
 export PATH=$SNAP/bin:$SNAP/usr/bin:/snap/bin:$PATH
-lxd_binary=$(which lxd)
-
-# Prefer snap installed lxd, if availible
-if [ "$lxd_binary" = "/snap/bin/lxd" ]; then
-    export LXD_DIR=$snap_lxd_socket
-elif [ "$lxd_binary" = "/usr/bin/lxd" ]; then
-    export LXD_DIR=$debian_socket
-fi
 export LD_LIBRARY_PATH
 LD_LIBRARY_PATH=$SNAP/lib:$SNAP/usr/lib/$(uname -p)-linux-gnu/
 

--- a/snap/wrappers/juju-wait
+++ b/snap/wrappers/juju-wait
@@ -1,0 +1,20 @@
+#!/bin/bash
+
+
+# Make sure we access our python and juju binaries first
+debian_socket=/var/lib/lxd
+snap_lxd_socket=/var/snap/lxd/common/lxd
+
+export PATH=$SNAP/bin:$SNAP/usr/bin:/snap/bin:$PATH
+lxd_binary=$(which lxd)
+
+# Prefer snap installed lxd, if availible
+if [ "$lxd_binary" = "/snap/bin/lxd" ]; then
+    export LXD_DIR=$snap_lxd_socket
+elif [ "$lxd_binary" = "/usr/bin/lxd" ]; then
+    export LXD_DIR=$debian_socket
+fi
+export LD_LIBRARY_PATH
+LD_LIBRARY_PATH=$SNAP/lib:$SNAP/usr/lib/$(uname -p)-linux-gnu/
+
+exec "$SNAP/bin/juju-wait" "$@"

